### DISCUSSION TITLE
WhisperAI crashes the docker-container

### DISCUSSION THREAD

**Discussion by Husky110** [open-webui/open-webui#4270] [General]
# Bug Report

## Description

**Bug Summary:**
Whenever I try to use the audio-recording feature my docker container crashes. 
Here's the container logs on the crash-event:
```
INFO  [apps.audio.main] file.content_type: audio/wav
Could not load library libcudnn_ops_infer.so.8. Error: libcudnn_ops_infer.so.8: cannot open shared object file: No such file or directory
Loading WEBUI_SECRET_KEY from file, not provided as an environment variable.Loading WEBUI_SECRET_KEY from .webui_secret_keyUSE_OLLAMA is set to true, starting ollama serve.CUDA is enabled, appending LD_LIBRARY_PATH to include torch/cudnn & cublas libraries.2024/08/01 05:20:28 routes.go:1099: INFO server config env="map[CUDA_VISIBLE_DEVICES: GPU_DEVICE_ORDINAL: HIP_VISIBLE_DEVICES: HSA_OVERRIDE_GFX_VERSION: OLLAMA_DEBUG:false OLLAMA_FLASH_ATTENTION:false OLLAMA_HOST:http://127.0.0.1:11434 OLLAMA_INTEL_GPU:false OLLAMA_KEEP_ALIVE:5m0s OLLAMA_LLM_LIBRARY: OLLAMA_MAX_LOADED_MODELS:0 OLLAMA_MAX_QUEUE:512 OLLAMA_MODELS:/root/.ollama/models OLLAMA_NOHISTORY:false OLLAMA_NOPRUNE:false OLLAMA_NUM_PARALLEL:0 OLLAMA_ORIGINS:[http://localhost https://localhost http://localhost:* https://localhost:* http://127.0.0.1 https://127.0.0.1 http://127.0.0.1:* https://127.0.0.1:* http://0.0.0.0 https://0.0.0.0 http://0.0.0.0:* https://0.0.0.0:* app://* file://* tauri://*] OLLAMA_RUNNERS_DIR: OLLAMA_SCHED_SPREAD:false OLLAMA_TMPDIR: ROCR_VISIBLE_DEVICES:]"time=2024-08-01T05:20:28.659Z level=INFO source=images.go:784 msg="total blobs: 53"time=2024-08-01T05:20:28.660Z level=INFO source=images.go:791 msg="total unused blobs removed: 0"time=2024-08-01T05:20:28.660Z level=INFO source=routes.go:1146 msg="Listening on 127.0.0.1:11434 (version 0.3.0)"time=2024-08-01T05:20:28.661Z level=WARN source=assets.go:94 msg="found running ollama" pid=10 path=/tmp/ollama1439945203time=2024-08-01T05:20:28.661Z level=WARN source=assets.go:94 msg="found running ollama" pid=13 path=/tmp/ollama2009279848time=2024-08-01T05:20:28.661Z level=WARN source=assets.go:94 msg="found running ollama" pid=10 path=/tmp/ollama2237447354time=2024-08-01T05:20:28.661Z level=WARN source=assets.go:94 msg="found running ollama" pid=10 path=/tmp/ollama3584355008time=2024-08-01T05:20:28.661Z level=WARN source=assets.go:100 msg="unable to cleanup stale tmpdir" path=/tmp/ollama868162921 error="remove /tmp/ollama868162921: directory not empty"time=2024-08-01T05:20:28.661Z level=INFO source=payload.go:30 msg="extracting embedded files" dir=/tmp/ollama161789335/runners
```
At the beginning it tried to download the whisperai-model, but was unable to do so:

```
INFO  [apps.audio.main] file.content_type: audio/wavWARNI [faster_whisper] An error occured while synchronizing the model Systran/faster-whisper-base from the Hugging Face Hub:Cannot find an appropriate cached snapshot folder for the specified revision on the local disk and outgoing traffic has been disabled. To enable repo look-ups and downloads online, pass 'local_files_only=False' as input.WARNI [faster_whisper] Trying to load the model directly from the local cache, if it exists.WARNI [apps.audio.main] WhisperModel initialization failed, attempting download with local_files_only=Falsebdb5e801-8e76-472e-923e-f780a5ca5cf0.wavINFO:     127.0.0.1:52588 - "GET /health HTTP/1.1" 200 OKCould not load library libcudnn_ops_infer.so.8. Error: libcudnn_ops_infer.so.8: cannot open shared object file: No such file or directory
```

**Steps to Reproduce:**
Run Open WebUI in docker an try to use the voice-recognition.

**Expected Behavior:**
Working...

**Actual Behavior:**
Container-crash

## Environment

- **Open WebUI Version:** 0.3.10
- **Ollama (if applicable):** 0.3.0

- **Operating System:** Ubuntu 22.04
- **Browser (if applicable):** Ungoogled Chromium

## Reproduction Details

**Confirmation:**

- [x ] I have read and followed all the instructions provided in the README.md.
- [x ] I am on the latest version of both Open WebUI and Ollama. (<--- well... the current DockerImage does not ship with the latest version... had to build it myself...)
- [ ] I have included the browser console logs. -> Nope... There are none...
- [x ] I have included the Docker container logs. -> see above

## Logs and Screenshots

**Docker Container Logs:**
See above

## Installation Method

Docker via Dockerfile setting "USE_CUDA" to true and "USE_OLLAMA" to true.
nvidia-smi-output on my system:
```
+---------------------------------------------------------------------------------------+
| NVIDIA-SMI 535.183.06             Driver Version: 535.183.06   CUDA Version: 12.2     |
|-----------------------------------------+----------------------+----------------------+
| GPU  Name                 Persistence-M | Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp   Perf          Pwr:Usage/Cap |         Memory-Usage | GPU-Util  Compute M. |
|                                         |                      |               MIG M. |
|=========================================+======================+======================|
|   0  NVIDIA GeForce RTX 3080 ...    On  | 00000000:01:00.0  On |                  N/A |
| N/A   46C    P8              20W / 115W |   7684MiB / 16384MiB |     21%      Default |
|                                         |                      |                  N/A |
+-----------------------------------------+----------------------+----------------------+

```

## Additional Information

I am not sure what actually happened there. I'm guessing that there was something wrong during the whisper-download and everything else is a subsequent error.. Maybe there need to be some commands run and everything is fine...

**Comment by nathanhere** [comment]
This also started happening to me sometime early this week or last week. I build docker images from each iteration, and when I exec'd into the open-webui container and inspected where the file "libcudnn_ops_infer.so.8" should be, **it was not present at all.**

Was wondering if something upstream changed in one of the dependencies?

Docker logs:

```
open-webui  | INFO:     172.18.0.1:46470 - "POST /audio/api/v1/speech HTTP/1.1" 200 OK
open-webui  | INFO  [apps.audio.main] file.content_type: audio/wav
open-webui  | Could not load library libcudnn_ops_infer.so.8. Error: libcudnn_ops_infer.so.8: cannot open shared object file: No such file or directory
```

Browser console:

![image](https://github.com/user-attachments/assets/467cdf53-000a-47b6-a4f4-b10144f953a6)

**Comment by nathanhere** [comment]
Doing some manual execution, I believe the exact line that is throwing the error is line 363 in backend/apps/audio/main.py:

Using iPython:
```
In [46]: segments, info = model.transcribe(file_path, beam_size=5)
Could not load library libcudnn_ops_infer.so.8. Error: libcudnn_ops_infer.so.8: cannot open shared object file: No such file or directory
Aborted (core dumped)
```

**Reply by CSAnetGmbH** [reply]
Same here, it looks like the Linux Contaner does not Support CUDA 12.1 anymore. The 3.8 Version works fine.

**Reply by nathanhere** [reply]
@CSAnetGmbH which 3.8 container were you referring to?

**Comment by nathanhere** [comment]
### A Solution
By adding this RUN instruction to the end of my dockerfile, Whisper no longer crashes the container (finally):
```bash
RUN apt update \
  && apt install -y wget \
  && wget https://developer.download.nvidia.com/compute/redist/cudnn/v8.2.1/cudnn-11.3-linux-x64-v8.2.1.32.tgz \
  && tar -xzvf cudnn-11.3-linux-x64-v8.2.1.32.tgz \
  && cp cuda/lib64/libcudnn* /usr/local/lib/python3.11/site-packages/nvidia/cudnn/lib \
  && rm -rf cuda \
  && rm cudnn*
```

### Explaination
This will download and extract the files in the screenshot, copy them over to " /usr/local/lib/python3.11/site-packages/nvidia/cudnn/lib" (because Whisper apparently needs more than just `libcudnn_ops_infer.so.8`, like `libcudnn_cnn_infer.so.8` and who knows what else) , then it removes all traces of the original download.

![image](https://github.com/user-attachments/assets/8175699d-9038-49e1-91da-16be3dc1e9ab)

**Reply by Mushy-Snugglebites-badonkadonk** [reply]
What comes installed in the open-webui:cuda version is:

```
:/usr/local/lib/python3.11/site-packages/nvidia/cudnn# ls lib/
__init__.py    libcudnn_adv.so.9                  libcudnn_engines_runtime_compiled.so.9  libcudnn_ops.so.9
__pycache__    libcudnn_cnn.so.9                  libcudnn_graph.so.9
libcudnn.so.9  libcudnn_engines_precompiled.so.9  libcudnn_heuristic.so.9
```

I wonder when this will be fixed or if this is even an issue with Open WebUI or if it's an issue with one of its dependencies. 
https://github.com/open-webui/open-webui/blob/main/backend/requirements.txt

**Reply by Mushy-Snugglebites-badonkadonk** [reply]
These are the only files you need to get Whisper (local) working again. Hope this helps!

![Screenshot 2024-09-04 at 03 36 38](https://github.com/user-attachments/assets/6eca2557-01f4-413d-ab01-bf59f7d5d0d8)

I'm guessing that the devs and maintainers of _faster-whisper_ will need to update their code to work with the newer version of libcudnn v.9, the one that came installed with Open WebUI. And I'm guessing because this crash is a dependency issue, that's why the original bug report was closed and moved to this discussion thread. 

@nathanhere Thank you so much for providing the solution!

**Reply by nathanhere** [reply]
@Mushy-Snugglebites-badonkadonk I’m so glad to know that my life struggles can help others :)

**Reply by Mushy-Snugglebites-badonkadonk** [reply]
@nathanhere I did some more tinkering and we don't even have to mess with any downloading. We can symlink the .so.8 files to the .so.9 files that _faster-whisper_ is looking for. 

```
#!/bin/bash

DESIRED_LIBS=("libcudnn.so.9" "libcudnn_adv.so.9" "libcudnn_cnn.so.9" "libcudnn_ops.so.9")
SOURCE_DIR="/usr/local/lib/python3.11/site-packages/nvidia/cudnn/lib"

# Greet the user
echo "Hello! Starting the CUDA library symlink creation process."

# Check if source directory exists
if [[ ! -d "$SOURCE_DIR" ]]; then
  echo "Source directory $SOURCE_DIR does not exist. Exiting."
  exit 1
fi

# Inform the user about the source directory being used
echo "Using source directory: $SOURCE_DIR"

for lib in "${DESIRED_LIBS[@]}"; do
  case $lib in
    "libcudnn.so.9")
      base_name="libcudnn.so.8"
      ;;
    "libcudnn_adv.so.9")
      base_name="libcudnn_adv_infer.so.8"
      ;;
    "libcudnn_cnn.so.9")
      base_name="libcudnn_cnn_infer.so.8"
      ;;
    "libcudnn_ops.so.9")
      base_name="libcudnn_ops_infer.so.8"
      ;;
  esac
  
  # Inform the user about the current operation
  echo "Creating symlink for $lib as $base_name"
  
  ln -sf "$SOURCE_DIR/$lib" "$SOURCE_DIR/$base_name"
  if [[ $? -eq 0 ]]; then
    echo "Successfully created symlink for $lib"
  else
    echo "Failed to create symlink for $lib"
  fi
done

# Inform the user that the process is complete
echo "CUDA library symlink creation process completed."

```

**Reply by Mushy-Snugglebites-badonkadonk** [reply]
Seems like this issue has been resolved in v.0.4.x as I noticed I am not having to run the symlink script anymore after updating the Docker container. 🥳
